link/tracing_test.go

```diff
@@ -47,6 +47,58 @@ func TestFreplace(t *testing.T) {
 	})
 }
 
+func TestFentryFexit(t *testing.T) {
+	testutils.SkipOnOldKernel(t, "5.5", "fentry")
+
+	testutils.Files(t, testutils.Glob(t, "../testdata/fentry_fexit-*.elf"), func(t *testing.T, file string) {
+		spec, err := ebpf.LoadCollectionSpec(file)
+		if err != nil {
+			t.Fatal("Can't parse ELF:", err)
+		}
+
+		if spec.ByteOrder != internal.NativeEndian {
+			return
+		}
+
+		target, err := ebpf.NewProgram(spec.Programs["target"])
+		testutils.SkipIfNotSupported(t, err)
+		if err != nil {
+			t.Fatal("Can't create target program:", err)
+		}
+		defer target.Close()
+
+		targetInfo, err := target.Info()
+		if err != nil {
+			t.Fatal("Can't get target program info:", err)
+		}
+
+		test := func(prog *ebpf.ProgramSpec, target *ebpf.Program) {
+			prog.AttachTarget = target
+			prog.AttachTo = targetInfo.Name
```

Comment:
If we set empty to `AttachTo`, ebpf will raise error. I added `debug.PrintStack()` to show the backtrace below:

```
github.com/cilium/ebpf.findTargetInProgram(0xc000522800?, {0xc00001c2d7, 0x2}, 0x59ab60?, 0x0?)
	/home/gray/src/github.com/cilium/ebpf/prog.go:1023 +0x186
github.com/cilium/ebpf.newProgramWithOptions(0xc00024d608, {0x59acc0?, 0xc0000725f0?, 0x0, 0x0?})
	/home/gray/src/github.com/cilium/ebpf/prog.go:299 +0xf0d
github.com/cilium/ebpf.(*collectionLoader).loadProgram(0xc0000147a0, {0x52d2ef, 0x8})
	/home/gray/src/github.com/cilium/ebpf/collection.go:521 +0x43c
github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign.func1({0x59d3e0?, 0x55a840}, {0x52d2ef, 0x8})
	/home/gray/src/github.com/cilium/ebpf/collection.go:294 +0x205
github.com/cilium/ebpf.assignValues({0x53be00, 0xc000088280?}, 0xc00024db40)
	/home/gray/src/github.com/cilium/ebpf/collection.go:827 +0x463
github.com/cilium/ebpf.(*CollectionSpec).LoadAndAssign(0xc000212a50, {0x53be00, 0xc000088280}, 0x7?)
	/home/gray/src/github.com/cilium/ebpf/collection.go:306 +0x1bc
```

The code there is to get target program's btf.Func, and `typeName` is whatever we set to `AttachTo`:

https://github.com/cilium/ebpf/blob/d1a52333f2c0fed085f8d742a5a3c164795d8492/prog.go#L1020-L1022

Reply:
Okay, finally got the time to take a closer look at this. I see, we need to resolve the funcinfo of the target prog. It seems like this PR doesn't fully implement fentry/fexit on user progs though? You basically need to use it like in the test case, manually loading individual progs from the CollectionSpec and fixing up ProgramSpec.AttachTo yourself? That's not ideal. Your patch extends the BTF lookup to deal with fentry/fexit, but that's only half the story.

Normally, you'd write something like `__section("fentry/target")` to point at the tracee symbol, where the string after the `/` is be used to populate the the AttachTo field for use in the BTF lookup. As I stated above, using Program.Info().Name for this will break for all targets with function names over 20 chars.

How does libbpf discern between user-defined symbols and kernel symbols?

Reply:
Thanks timo! You are correct, Program.Info().Name has string length limitation, this PR doesn't take care of it. 
I just submit a draft PR to provide an API to get full program name: https://github.com/cilium/ebpf/pull/1129
Do you think it's an acceptable solution for it?

Reply:
I took a look at your PR, I think we might not even need new API? Like Timo says, change the section names in `fentry_fexit.c`. Then you always have the correct AttachTo populated by the ELF loader, the only thing the test needs to do is set `AttachTarget`.